File: mc-clock.js
#!/usr/bin/node
const net = require('node:net');
const client = net.createConnection({ port: 7070 }, () => {
  // 'connect' listener.
  console.log('connected to server!');
});


client.on('data', (data) => {
  //console.log(data.toString());
});


client.on('end', () => {
  console.log('disconnected from server');
  process.exit(0)
}); 


setInterval( updateClock, 1000 );


function updateClock() {
  const d = new Date();
  const str = pad(d.getHours())+":"+pad(d.getMinutes())+":"+pad(d.getSeconds());
  const pic = strToPic(str);
  pic.forEach( (line)=>console.log( line.join( '' )))
	console.log()
  setPic( 0, 150, 500, pic );
}
function setPic( x,y,z, pic ) {
  for( let yo=0; yo<pic.length; yo++ ) {
    for( let xo=0; xo<pic[yo].length; xo++ ) {
      if( pic[yo][xo] == "#" ) {
        setBlock(x+xo, y+pic.length-1-yo, z, "minecraft:lime_wool" );
      }
      if( pic[yo][xo] == " " ) {
        setBlock(x+xo, y+pic.length-1-yo, z, "minecraft:stone" );
      }
    }
  }
}

function strToPic(str) {
  const template = `
0__ 1__ 2__ 3__ 4__ 5__ 6__ 7__ 8__ 9__ :__ X
###   # ### ### # # ### ### ### ### ###     X
# #   #   #   # # # #   #     # # # # #  #  X
# #   # ### ### ### ### ###   # ### ###     X
# #   # #     #   #   # # #   # # #   #  #  X
###   # ### ###   # ### ###   # ###   #     X
`;

  const lines = template.split( /\n/ );
  const index = {};
  for( var i=0; i<lines[1].length; i+=4 ) {
    index[""+lines[1][i]] = i;
  }
  
  var outLines = [[' '],[' '],[' '],[' '],[' ']];
  for( i=0; i<str.length;i++ ) {
    var off = index[str[i]];
    for( row=0;row<5;++row ) {
      const rowText = lines[row+2];
      for( col=0;col<4;++col ) {
        outLines[row].push(rowText.substr(off+col,1));
      }
   }
  } 
  //outLines.forEach( (item)=>{ console.log( item.join( "" )); });
  return outLines;
}

// zero pad 2 digit numbers
function pad(n) {
  return ('0'+n).slice(-2);
}

function setBlock(x,y,z,type ) {
  client.write( `setBlock ${x} ${y} ${z} ${type}\n` )
}

File: mc-clock.py
#!/usr/bin/env python3
import socket
import time
from datetime import datetime

# Connect to the server
client = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
client.connect(('localhost', 7070))
print("connected to server!")

# Handle receiving data from the server
def receive_data():
    try:
        while True:
            data = client.recv(1024)
            if not data:
                break
    except Exception as e:
        print(f"Error receiving data: {e}")
    finally:
        print("disconnected from server")
        client.close()

# Function to send the block data
def set_block(x, y, z, block_type):
    command = f"setBlock {x} {y} {z} {block_type}\n"
    client.sendall(command.encode())

# Zero pad 2-digit numbers
def pad(n):
    return str(n).zfill(2)

# Function to convert string to picture-like format
def str_to_pic(string):
    template = """
0__ 1__ 2__ 3__ 4__ 5__ 6__ 7__ 8__ 9__ :__ X
###   # ### ### # # ### ### ### ### ###     X
# #   #   #   # # # #   #     # # # # #  #  X
# #   # ### ### ### ### ###   # ### ###     X
# #   # #     #   #   # # #   # # #   #  #  X
###   # ### ###   # ### ###   # ###   #     X
"""
    lines = template.strip().split('\n')
    index = {lines[0][i]: i for i in range(0, len(lines[0]), 4)}

    out_lines = [[' '] for _ in range(5)]
    for char in string:
        off = index.get(char, 0)
        for row in range(5):
            row_text = lines[row + 1]
            for col in range(4):
                out_lines[row].append(row_text[off + col])
    return out_lines

# Function to set picture blocks
def set_pic(x, y, z, pic):
    for yo, row in enumerate(pic):
        for xo, val in enumerate(row):
            if val == "#":
                set_block(x + xo, y + len(pic) - 1 - yo, z, "minecraft:lime_wool")
            elif val == " ":
                set_block(x + xo, y + len(pic) - 1 - yo, z, "minecraft:green_wool")

# Function to update the clock
def update_clock():
    while True:
        now = datetime.now()
        clock_str = f"{pad(now.hour)}:{pad(now.minute)}:{pad(now.second)}"
        pic = str_to_pic(clock_str)
        for line in pic:
            print("".join(line))
        print()
        set_pic(0, 150, 500, pic)
        time.sleep(1)

# Start the clock update loop
try:
    update_clock()
except KeyboardInterrupt:
    print("Clock stopped by user.")
    client.close()

